src/test/cls_rgw/test_cls_rgw.cc

```diff
@@ -1340,3 +1340,133 @@ TEST_F(cls_rgw, index_racing_removes)
+                           std::list<rgw_cls_bi_entry> *entries, bool *is_truncated)
+{
+  int ret = cls_rgw_bi_list(ioctx, oid, "", "", 100, entries, is_truncated, true);
+  if (ret < 0) {
+    return ret;
+  }
+  return 0;
+}
+
+TEST_F(cls_rgw, reshardlog_list)
+{
+  string bucket_oid = str_int("reshard", 0);
+
+  ObjectWriteOperation op;
+  cls_rgw_bucket_init_index(op);
+  ASSERT_EQ(0, ioctx.operate(bucket_oid, &op));
+
+  cls_rgw_obj_key obj1 = str_int("obj1", 0);
+  string tag = str_int("tag-prepare", 0);
+  string loc = str_int("loc", 0);
+  index_prepare(ioctx, bucket_oid, CLS_RGW_OP_ADD, tag, obj1, loc);
+  rgw_bucket_dir_entry_meta meta;
+  index_complete(ioctx, bucket_oid, CLS_RGW_OP_ADD, tag, 1, obj1, meta);
+
+  // do not record logs
+  bool is_truncated = false;
+  std::list<rgw_cls_bi_entry> entries;
+  ASSERT_EQ(0, reshardlog_list(ioctx, bucket_oid, &entries, &is_truncated));
+  ASSERT_FALSE(is_truncated);
+  ASSERT_EQ(0u, entries.size());
+
+  // set reshard status to IN_LOGRECORD
+  cls_rgw_bucket_instance_entry entry;
+  entry.reshard_status = cls_rgw_reshard_status::IN_LOGRECORD;
+  set_reshard_status(ioctx, bucket_oid, entry);
+
+  // record a log in prepare
+  cls_rgw_obj_key obj2 = str_int("obj2", 0);
+  entries.clear();
+  index_prepare(ioctx, bucket_oid, CLS_RGW_OP_ADD, tag, obj2, loc);
+  ASSERT_EQ(0, reshardlog_list(ioctx, bucket_oid, &entries, &is_truncated));
+  ASSERT_FALSE(is_truncated);
+  ASSERT_EQ(1u, entries.size());
+
+  // overwrite the log writen in prepare
+  entries.clear();
+  index_complete(ioctx, bucket_oid, CLS_RGW_OP_ADD, tag, 1, obj2, meta);
+  ASSERT_EQ(0, reshardlog_list(ioctx, bucket_oid, &entries, &is_truncated));
+  ASSERT_FALSE(is_truncated);
+  ASSERT_EQ(1u, entries.size());
+
+  // record a log in deleting obj
+  entries.clear();
+  index_prepare(ioctx, bucket_oid, CLS_RGW_OP_DEL, tag, obj1, loc);
+  index_complete(ioctx, bucket_oid, CLS_RGW_OP_DEL, tag, 1, obj1, meta);
+  ASSERT_EQ(0, reshardlog_list(ioctx, bucket_oid, &entries, &is_truncated));
+  ASSERT_FALSE(is_truncated);
+  ASSERT_EQ(2u, entries.size());
+
+  // overwrite the log writen
+  entries.clear();
+  index_prepare(ioctx, bucket_oid, CLS_RGW_OP_DEL, tag, obj2, loc);
+  index_complete(ioctx, bucket_oid, CLS_RGW_OP_DEL, tag, 1, obj2, meta);
+  ASSERT_EQ(0, reshardlog_list(ioctx, bucket_oid, &entries, &is_truncated));
+  ASSERT_FALSE(is_truncated);
+  ASSERT_EQ(2u, entries.size());
+}
+
+void reshardlog_entries(librados::IoCtx& ioctx, const std::string& oid, uint32_t num_entries)
+{
+  map<int, struct rgw_cls_list_ret> results;
+  map<int, string> oids;
+  oids[0] = oid;
+  ASSERT_EQ(0, CLSRGWIssueGetDirHeader(ioctx, oids, results, 8)());
+
+  uint32_t entries = 0;
+  map<int, struct rgw_cls_list_ret>::iterator iter = results.begin();
+  for (; iter != results.end(); ++iter) {
+    entries += (iter->second).dir.header.reshardlog_entries;
+  }
+  ASSERT_EQ(entries, num_entries);
```

Comment:
i ran this through the rgw teuthology suite in https://pulpito.ceph.com/cbodley-2024-08-20_14:00:40-rgw-pr-56597-distro-default-smithi/. that caught some ceph_test_cls_rgw failures here:
```
2024-08-20T14:49:13.440 INFO:tasks.workunit.client.0.smithi130.stdout:[ RUN      ] cls_rgw.reshardlog_list
2024-08-20T14:49:13.463 INFO:tasks.workunit.client.0.smithi130.stdout:[       OK ] cls_rgw.reshardlog_list (23 ms)
2024-08-20T14:49:13.463 INFO:tasks.workunit.client.0.smithi130.stdout:[ RUN      ] cls_rgw.reshardlog_num
2024-08-20T14:49:13.473 INFO:tasks.workunit.client.0.smithi130.stdout:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-5172-ge2fb2c63/rpm/el9/BUILD/ceph-19.0.0-5172-ge2fb2c63/src/test/cls_rgw/test_cls_rgw.cc:1434: Failure
2024-08-20T14:49:13.473 INFO:tasks.workunit.client.0.smithi130.stdout:Expected equality of these values:
2024-08-20T14:49:13.473 INFO:tasks.workunit.client.0.smithi130.stdout:  entries
2024-08-20T14:49:13.473 INFO:tasks.workunit.client.0.smithi130.stdout:    Which is: 2
2024-08-20T14:49:13.474 INFO:tasks.workunit.client.0.smithi130.stdout:  num_entries
2024-08-20T14:49:13.474 INFO:tasks.workunit.client.0.smithi130.stdout:    Which is: 1
2024-08-20T14:49:13.474 INFO:tasks.workunit.client.0.smithi130.stdout:[  FAILED  ] cls_rgw.reshardlog_num (10 ms)
```

Reply:
it should be easy to reproduce these locally with vstart:
```
~/ceph/build $ OSD=1 MON=1 RGW=0 MGR=0 MDS=0 ../src/vstart.sh -n -d
~/ceph/build $ bin/ceph_test_cls_rgw --gtest_filter=cls_rgw.reshardlog_num
```

Reply:
no, ceph_test_cls_rgw runs directly against the osd

Reply:
done. I did not increase the epoch, so `CLS_RGW_OP_DELETE` is changed to `CLS_RGW_OP_CANCEL`

Reply:
still seeing this failure in testing:
```
2024-09-01T00:16:21.408 INFO:tasks.workunit.client.0.smithi003.stdout:[ RUN      ] cls_rgw.reshardlog_num
2024-09-01T00:16:21.438 INFO:tasks.workunit.client.0.smithi003.stdout:./src/test/cls_rgw/test_cls_rgw.cc:1446: Failure
2024-09-01T00:16:21.438 INFO:tasks.workunit.client.0.smithi003.stdout:Expected equality of these values:
2024-09-01T00:16:21.438 INFO:tasks.workunit.client.0.smithi003.stdout:  entries
2024-09-01T00:16:21.438 INFO:tasks.workunit.client.0.smithi003.stdout:    Which is: 2
2024-09-01T00:16:21.439 INFO:tasks.workunit.client.0.smithi003.stdout:  num_entries
2024-09-01T00:16:21.439 INFO:tasks.workunit.client.0.smithi003.stdout:    Which is: 1
2024-09-01T00:16:21.439 INFO:tasks.workunit.client.0.smithi003.stdout:[  FAILED  ] cls_rgw.reshardlog_num (31 ms)
```
(sorry the line number doesn't match up, i had rebased your pr for testing)

Reply:
@cbodley  Yep, I tested it successfully at  the first, then accounted deleting for reshardlog entries according your advice. But forgot to change this testing place.